### Practical 21

### Comparision between accuracy of K-NN & Logistic Regression and method is suitable for the given dataset.

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

####                                                                                  K-NN Algorithm   

In [2]:
data = pd.read_csv("D:/ML Dataset/User_Data.csv")
data.head(5)

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [3]:
print(data.describe())

            User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000


In [4]:
data["Purchased"].value_counts()

0    257
1    143
Name: Purchased, dtype: int64

In [5]:
data.columns

Index(['User ID', 'Gender', 'Age', 'EstimatedSalary', 'Purchased'], dtype='object')

In [6]:
x = data[['Age', 'EstimatedSalary']].values
print("x :",x[0:5])
y = data[ 'Purchased'].values
print("y :",y[0:5])

x : [[   19 19000]
 [   35 20000]
 [   26 43000]
 [   27 57000]
 [   19 76000]]
y : [0 0 0 0 0]


In [7]:
x = preprocessing.StandardScaler().fit(x).transform(x.astype(float))
# in this step we are using anonymous objects. linking objects without makig them refer is called anonymous objects
x[0:10]

array([[-1.78179743, -1.49004624],
       [-0.25358736, -1.46068138],
       [-1.11320552, -0.78528968],
       [-1.01769239, -0.37418169],
       [-1.78179743,  0.18375059],
       [-1.01769239, -0.34481683],
       [-1.01769239,  0.41866944],
       [-0.54012675,  2.35674998],
       [-1.20871865, -1.07893824],
       [-0.25358736, -0.13926283]])

In [8]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [9]:
k = 4
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 4 is :", metrics.accuracy_score(y_test, y_pred))

Accuracy of model at K = 4 is : 0.95


In [10]:
k = 5
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 5 is :", metrics.accuracy_score(y_test, y_pred))

Accuracy of model at K = 5 is : 0.95


In [11]:
k = 6
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 6 is :", metrics.accuracy_score(y_test, y_pred))

Accuracy of model at K = 6 is : 0.95


In [12]:
k = 7
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 7 is :", metrics.accuracy_score(y_test, y_pred))

Accuracy of model at K = 7 is : 0.95


In [13]:
k = 8
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 8 is :", metrics.accuracy_score(y_test, y_pred))

Accuracy of model at K = 8 is : 0.95


In [14]:
k = 8
knnClassifier = KNeighborsClassifier(n_neighbors = k, metric = "minkowski", p = 2).fit(x_train, y_train)
y_pred = knnClassifier.predict(x_test)
print("Accuracy of model at K = 8 is :", metrics.accuracy_score(y_test, y_pred))
confusion_matrix = metrics.confusion_matrix(y_test, y_pred)
print(confusion_matrix)

Accuracy of model at K = 8 is : 0.95
[[55  3]
 [ 1 21]]


In [15]:
print(metrics.classification_report(y_test, y_pred))
print("Accuracy of model at K = 8 is :", metrics.accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.95      0.96        58
           1       0.88      0.95      0.91        22

    accuracy                           0.95        80
   macro avg       0.93      0.95      0.94        80
weighted avg       0.95      0.95      0.95        80

Accuracy of model at K = 8 is : 0.95


####   Logistic Regression 

In [16]:
data.head(5)

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [17]:
x = data.iloc[:,[2,3]].values
y = data.iloc[:,4].values
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [18]:
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)
print(x_train[0:5])

[[ 1.92295008  2.14601566]
 [ 2.02016082  0.3787193 ]
 [-1.3822153  -0.4324987 ]
 [-1.18779381 -1.01194013]
 [ 1.92295008 -0.92502392]]


In [19]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(x_train,y_train)

LogisticRegression(random_state=0)

In [20]:
y_pred = classifier.predict(x_test)
y_pred

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1], dtype=int64)

In [21]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)
print("Confusion Matrix :\n", cm)

Confusion Matrix :
 [[57  1]
 [ 5 17]]


In [22]:
print('Accuracy :', metrics.accuracy_score(y_test,y_pred))
print('Precision :', metrics.precision_score(y_test,y_pred))
print('Recall :', metrics.recall_score(y_test,y_pred))

Accuracy : 0.925
Precision : 0.9444444444444444
Recall : 0.7727272727272727


In [23]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.92      0.98      0.95        58
           1       0.94      0.77      0.85        22

    accuracy                           0.93        80
   macro avg       0.93      0.88      0.90        80
weighted avg       0.93      0.93      0.92        80



####    Conclusion

In K-NN the accuracy is 95%
But in Logistic Regression the accuracy is 92.5%
So we can conclude that for this dataset K-NN algorithm is prefered.